cachi2/core/package_managers/yarn_classic/workspaces.py

```diff
@@ -0,0 +1,114 @@
+import json
+from contextlib import suppress
+from itertools import chain
+from pathlib import Path
+from typing import Any, Iterable
+
+import pydantic
+
+from cachi2.core.errors import PackageRejected
+from cachi2.core.models.input import YarnClassicPackageInput
+from cachi2.core.rooted_path import RootedPath
+
+
+class Workspace(pydantic.BaseModel):
+    """Workspace model."""
+
+    path: Path  # path to a workspace.
+    package_contents: dict  # package data extracted from path/"package.json".
+    # package reference for potential nested workspace extraction:
+    package: YarnClassicPackageInput
+
+
+def ensure_no_path_leads_out(
+    paths: Iterable[Path],
+    source_dir: RootedPath,
+) -> None:
+    """Ensure no path leads out of source directory.
+
+    Raises an exception when any path is not relative to source directory.
+    Does nothing when path does not exist in the file system.
+    """
+    for path in paths:
+        if not path.is_relative_to(source_dir.path):
+            raise PackageRejected(
+                f"Found a workspace path which is not relative to package: {path}",
+                solution=(
+                    "Avoid using packages which try to access your filesystem "
+                    "outside of package directory."
+                ),
+            )
```

Comment:
Would the helper need more than what RootedPath already provides? I might be missing some nuance here

https://github.com/containerbuildsystem/cachi2/blob/c94a1a24c1777036bbed121281c41be1b8f156f1/cachi2/core/rooted_path.py#L111

Reply:
Good catch! Unless we want to traverse the hierarchy, because RootedPath can only handle the immediate descendants, but not ancestors up to the real "root".

Reply:
Done.